### PR TITLE
Modify ASP.NET Core instrumentation to take advantage of sampling decision

### PR DESCRIPTION
**Pull request by utpilla**
## Changes
- Extends the optimization done in #1894 to ASP.NET Core instrumentation
- Added a unit test for context propagation even when activity.IsAllDataRequested is false

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 1]
```diff
@@ -57,29 +57,29 @@ public HttpInListener(string name, AspNetCoreInstrumentationOptions options)
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Reliability", "CA2000:Dispose objects before losing scope", Justification = "The objects should not be disposed.")]
         public override void OnStartActivity(Activity activity, object payload)
         {
-            _ = this.startContextFetcher.TryFetch(payload, out HttpContext context);
-            if (context == null)
+            // The overall flow of what HttpClient library does is as below:
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:60]:
```suggestion
            // The overall flow of what AspNetCore library does is as below:
```

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 2]
```diff
@@ -57,29 +57,29 @@ public HttpInListener(string name, AspNetCoreInstrumentationOptions options)
...
+            _ = this.startContextFetcher.TryFetch(payload, out HttpContext context);
+            if (context == null)
             {
-                AspNetCoreInstrumentationEventSource.Log.RequestFilterException(ex);
-                activity.IsAllDataRequested = false;
+                AspNetCoreInstrumentationEventSource.Log.NullPayload(nameof(HttpInListener), nameof(this.OnStartActivity));
                 return;
             }
 
+            // Enusre context propagation irrespective of sampling decision
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:82]:
```suggestion
            // Ensure context propagation irrespective of sampling decision
```

Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:82]:
```suggestion
            // Enusre context extraction irrespective of sampling decision
```

File: test/TestApp.AspNetCore.2.1/Controllers/ChildActivityController.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            activity.Stop();
+            return result;
+        }
+
+        [Route("api/GetChildActivityBaggageContext")]
+        public IReadOnlyDictionary<string, string> GetChildActivityBaggageContext()
+        {
+            var activity = new Activity("ActivityInsideHttpRequest");
+            activity.Start();
+            var result = Baggage.Current.GetBaggage();
```
Comment (cijothomas) [inline test/TestApp.AspNetCore.2.1/Controllers/ChildActivityController.cs:44]: is activity start/stop here relevant?

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs [hunk 4]
```diff
@@ -383,8 +391,62 @@ void ConfigureTestServices(IServiceCollection services)
...
+
+                Assert.Equal(expectedTraceId.ToString(), childActivityTraceContext["TraceId"]);
+                Assert.Equal(expectedTraceState, childActivityTraceContext["TraceState"]);
+                Assert.NotEqual(expectedParentSpanId.ToString(), childActivityTraceContext["ParentSpanId"]); // there is a new activity created in instrumentation therefore the ParentSpanId is different that what is provided in the headers
+
+                // Test Baggage Context Propagation
+                request = new HttpRequestMessage(HttpMethod.Get, "/api/GetChildActivityBaggageContext");
+                request.Headers.Add("Baggage", "key1=value1,key2=value2");
+                response = await client.SendAsync(request);
+                var childActivityBaggageContext = JsonConvert.DeserializeObject<IReadOnlyDictionary<string, string>>(response.Content.ReadAsStringAsync().Result);
```
Comment (cijothomas) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:438]: you can avoid the additional json parsing, by making the test app return 2 strings contaning traceparent and tracestate (activity.Id, activity.TraceState) respectively.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 5]
```diff
@@ -110,11 +110,29 @@ public override void OnStartActivity(Activity activity, object payload)
...
-            ActivityInstrumentationHelper.SetActivitySourceProperty(activity, ActivitySource);
-            ActivityInstrumentationHelper.SetKindProperty(activity, ActivityKind.Server);
-
+            // enrich Activity from payload only if sampling decision
+            // is favorable.
             if (activity.IsAllDataRequested)
             {
+                try
+                {
+                    if (this.options.Filter?.Invoke(context) == false)
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:119]: if not exists, you can add a test validating that Filter is only called if sampling is favorable.. Same for Enrich.

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs [hunk 6]
```diff
@@ -383,8 +391,62 @@ void ConfigureTestServices(IServiceCollection services)
...
+                        .AddProcessor(activityProcessor.Object)
+                        .Build();
+                    })))
+            {
+                using var client = testFactory.CreateClient();
+
+                // Test TraceContext Propagation
+                var expectedTraceState = "rojo=1,congo=2";
+                var request = new HttpRequestMessage(HttpMethod.Get, "/api/GetChildActivityTraceContext");
+                request.Headers.Add("traceparent", $"00-{expectedTraceId}-{expectedParentSpanId}-01");
```
Comment (cijothomas) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:422]: portions of this test would pass by default, as asp.net creates activity from traceparent always, irrespective of sampling in Otel.
To ensure thorough test, use a custom propagator (with some header like Foo), and validate the scenarios. Also add cases where sending traceparent with "00" and "01" as sampling flags

File: test/TestApp.AspNetCore.2.1/Controllers/ChildActivityController.cs (hunk 3 above)
Comment (utpilla) [reply test/TestApp.AspNetCore.2.1/Controllers/ChildActivityController.cs:42]: Removed this.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs (hunk 5 above)
Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:119]: Added a unit test for testing both Filter and Enrich are only called when sampled.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 7]
```diff
@@ -57,29 +57,29 @@ public HttpInListener(string name, AspNetCoreInstrumentationOptions options)
...
+            // Activity.Start()
+            // DiagnosticSource.WriteEvent("Start", payload)
+            // DiagnosticSource.WriteEvent("Stop", payload)
+            // Activity.Stop()
+
+            // This method is in the WriteEvent("Start", payload) path.
+            // By this time, samplers have already run and
+            // activity.IsAllDataRequested populated accordingly.
+
+            if (Sdk.SuppressInstrumentation)
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:70]: For another PR: Confirm that there is a unittest validating that this instrumentation respects Suppress Instrumentation.

**Review by cijothomas — APPROVED** [review]

**Review by cijothomas — APPROVED** [review]
Lets add to changelog that FIlter/Enrich is only called if AllDataRequested is true.

**Pull request merged**